File: Tecniques de Programació Avançades/Practica 2 - Aleix Carrillo i Oriol Poblet/src/main/java/BigGroup.scala
import scala.collection.JavaConverters._
import scala.collection.convert.ImplicitConversions.`seq AsJavaList`

class BigGroup[U, V](groupSize: Int) extends PolicyManager[U, V] {
  /**
   * Método que devuelve un invoker que cumple con los requisitos de memoria
   *
   * @param invokers       Lista de invokers
   * @param requiredMemory Memoria requerida para la ejecución
   * @return Invoker que cumple con los requisitos de memoria
   */
  override def valorsInvoker(invokers: List[Invoker[U, V]], requiredMemory: Int): Invoker[U, V] = {
    var remainingMemory = requiredMemory
    var j = 0

    while (remainingMemory > 0) {
      val currentGroupSize = Math.min(groupSize, remainingMemory)
      val currentGroup = new java.util.ArrayList[Invoker[U, V]]

      for (i <- 0 until currentGroupSize) {
        if (j >= invokers.size) {
          j = 0
        }

        val invoker = invokers.get(j)
        currentGroup.add(invoker)
        j += 1
      }

      for (invoker <- currentGroup.asScala) {
        if (invoker.estaDisponible(requiredMemory)) {
          return invoker // Devuelve el primer invoker disponible
        }
      }
      remainingMemory -= currentGroupSize
    }
    throw new NoSuchElementException("No hay Invokers disponibles")
  }
}

File: Tecniques de Programació Avançades/Practica 1/src/BigGroup.java
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public class BigGroup implements PolicyManager {
    private int groupSize;

    /**
     * Politica Big Group: Envia a los invokers según el tamaño de los grupos
     * @param groupSize Tamaño de los grupo
     */

    public BigGroup(int groupSize) {
        this.groupSize = groupSize;
    }

    /**
     * Método que devuelve un invoker que cumple con los requisitos de memoria
     * @param invokers lista de invokers
     * @param requiredMemory memoria necesitada para la ejecución
     * @return invokerDisponible un invoker en uso
     */

    @Override
    public Invoker valorsInvoker(List<Invoker> invokers, int requiredMemory) {
        int remainingMemory = requiredMemory;
        int j = 0;

        while (remainingMemory > 0) {
            int currentGroupSize = Math.min(groupSize, remainingMemory);
            List<Invoker> currentGroup = new ArrayList<>();

            for (int i = 0; i < currentGroupSize; i++) {
                if (j >= invokers.size()) {
                    j = 0;
                }

                Invoker invoker = invokers.get(j++);
                currentGroup.add(invoker);
            }

            for (Invoker invoker : currentGroup) {
                if (invoker.estaDisponible(requiredMemory)) {
                    return invoker;  // Return the first available invoker
                }
            }

            remainingMemory -= currentGroupSize;
        }

        throw new NoSuchElementException("No hay Invokers disponibles");
    }
}
